core/controllers/reader.py

```diff
@@ -1371,6 +1371,172 @@ def get(self):
+            synced_exp_user_data = None
+            # If the latest exploration version is ahead of the most recently
+            # interacted exploration version.
+            if (
+                logged_out_user_data.most_recently_reached_checkpoint_exp_version is not None and # pylint: disable=line-too-long
+                logged_out_user_data.most_recently_reached_checkpoint_exp_version < exploration.version # pylint: disable=line-too-long
+            ):
+                synced_exp_user_data = (
+                    exp_services.sync_learner_checkpoint_progress_with_current_exp_version( # pylint: disable=line-too-long
+                        logged_out_user_data.exploration_id, unique_progress_url_id)) # pylint: disable=line-too-long
+            else:
+                synced_exp_user_data = logged_out_user_data
+
+            furthest_reached_checkpoint_exp_version = (
+                synced_exp_user_data.furthest_reached_checkpoint_exp_version)
+            furthest_reached_checkpoint_state_name = (
+                synced_exp_user_data.furthest_reached_checkpoint_state_name)
+            most_recently_reached_checkpoint_exp_version = (
+                synced_exp_user_data
+                    .most_recently_reached_checkpoint_exp_version)
+            most_recently_reached_checkpoint_state_name = (
+                synced_exp_user_data
+                    .most_recently_reached_checkpoint_state_name)
+
+        self.values.update({
+            'exploration_id': (
+                logged_out_user_data.exploration_id),
+            'furthest_reached_checkpoint_exp_version': (
+                furthest_reached_checkpoint_exp_version),
+            'furthest_reached_checkpoint_state_name': (
+                furthest_reached_checkpoint_state_name),
+            'most_recently_reached_checkpoint_exp_version': (
+                most_recently_reached_checkpoint_exp_version),
+            'most_recently_reached_checkpoint_state_name': (
+                most_recently_reached_checkpoint_state_name),
+        })
+
+        self.render_json(self.values)
+
+
+class SaveTransientCheckpointProgressHandler(base.BaseHandler):
+    """Responsible for storing progress of a logged-out learner."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+                'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        },
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'POST': {
+            'most_recently_reached_checkpoint_state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                }
+            },
+            'most_recently_reached_checkpoint_exp_version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            }
+        },
+        'PUT': {
+            'exploration_id': {
```

Comment:
Why add exploration id in PUT request schema when it's not there in the POST request ?

Reply:
Fixed.